### ISSUE TITLE
createField: Props assigned to WrappedComponent are not propagated properly

### ISSUE THREAD

**Issue by kettanaito** [kettanaito/react-advanced-form#143]
## What
Direct props are not propagated properly to `WrappedComponent` inside `createField.Field.render`.

## Why
Right now only `directProps.placeholder` and `directProps.style` props are passed properly. I remember there was an issue when destructing `directProps` entirely.

Needs investigation and quickfix.

**Comment by kettanaito** [comment]
After some investigation I see that the long discovered issue (#38) creates obstacles for the further development.

I will try to refine how props are handled, and there is a high possibility we will get rid of `connectField()` and just utilize `createField()` for styling purposes as well (or vice versa). Speaking of namespace I prefer `connectField()` for both styled fields and custom/third-party fields.

**Comment by kettanaito** [comment]
## Update
The list of changes introduced in #145:

* Native fields moved into the dedicated package
* Removed `connectField()` HOC used for custom styling
* Refined `createField()` to expect props destruction on any default form element (input, select, etc.)
* Refined props management on `createField()` level: separated the props into three distinct groups:
  * Props passed directly to the `CustomInput`
  * Context props of internal field record - `contextProps`
  * Props accepted by generic `Field` class
* Refined field registration logic. Fields are now registered with a defined record object, without explicit props destruction.
* Changed `createField.mapPropsToField` method to be called *after* the registration record is composed.
* `createField.mapPropsToField` now accepts `fieldRecord` Object to modify in case of need.
* `createField.mapPropsToField` now accepts **argument props**.
* Introduced basic fields set using `createField()` for testing purposes. Prototyping fields are going to be implemented in the dedicated package and include more than just plain form elements (error display, etc.).

**Comment by kettanaito** [comment]
The changes have been merged and published under [`0.17.0`](https://github.com/kettanaito/react-advanced-form/commit/9ec2f0ef473ffcde53690b7510868953c580b459).

I will provide the production testing and close the related issues if the testing passes.

**Comment by kettanaito** [comment]
## Issues
The issue I see rising from this change is that:

1. Now you need to define a lot of logic for custom fields. For example, to create a custom Radio, or just to style it, you now need to repeat all that is written in the prototyping `Radio` component.
1. No easy way to have an `onChange` handler when you *don't want* to make the field controlled, but just react to the change. Now you need to override `onChange` to your custom method, inside which you should call `this.props.handleFieldChange()`. This would work OK for third-party fields integration, but for basic fields it's too much of code for a simple event listener.

**Comment by kettanaito** [comment]
## Update
1. Event handlers issue was caused because I forgot to add `onChange` and other primary event handlers to the `immutableFieldRecord` Object. Now it works nicely without needing to explicitly write the logic.
1. Options repetition was solved by introducing `fieldPresets`. Those are Objects containing pre-defined options for `createField()` HOC to create a proper field record/etc.

**Comment by kettanaito** [comment]
After thinking about the remaining issue, it appears that there is no clear way to differentiate which props passed to `<CustomInput>` are meant for `<input>` itself, and which are passed as custom props which contribute to some internal logic of `CustomInput`.

I think that is a responsibility of developer to map such things as `placeholder` to the actual `<input>` element. I would be against any assumptions on the library's side on how to categorize the props.

What do you think, @Vidlec?

**Comment by kettanaito** [comment]
Another solution for this would be to enforce certain props like `placeholder` using field presets.

### Pros:
* Most common field props are mapped automatically

### Cons:
* Using presets would be mandatory for all the field types (is it really a drawback?)

**Comment by kettanaito** [comment]
@Vidlec, what do you think about the comment above?

**Comment by Vidlec** [comment]
Sounds interesting... 
So do I understand correctly that:
Solution 1) Let Developer handle it - its his responsibility to pass placeholder (or any prop) to input directly
Solution 2) Map this props in field presets - this mean hardcode most of them?
-------------------------------------------------------------------------------------------------

What about solution 3) introduce prop directProps and this is passes on input directly
```js
const directProps = {
  placeholder: 'blah'
}
```

**Comment by kettanaito** [comment]
@Vidlec, yes, you understood the options correctly.

I wouldn't go with the third option because this way the end usage of custom fields becomes unclear. Just imagine:
```jsx
<Input directProps={{ placeholder: 'foo' }} />
```
Direct props for what? From the readability point of view it becomes confusing. From the declaration point of view - as well, since now you need to pass those props like an Object.

**Comment by kettanaito** [comment]
I will add some default mapping of essential props for different form elements.

**Comment by kettanaito** [comment]
The essential form element's props are not automatically mapped to the actual form element using `enforceProps` option of `createField`.

Field presets were updated. Now each field type required to use a preset. Otherwise, the mapping must be performed by the developer.

Merged under [`0.20.2`](https://github.com/kettanaito/react-advanced-form/commit/23196329c77d53ff3a02ae5ca90523af7cd260ec). Closing.
